// @mockoon/commons@2.14.0/dist/cjs/models/route.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Methods = exports.RouteType = exports.ResponseMode = exports.BodyTypes = void 0;
var BodyTypes;
(function (BodyTypes) {
    BodyTypes["INLINE"] = "INLINE";
    BodyTypes["FILE"] = "FILE";
    BodyTypes["DATABUCKET"] = "DATABUCKET";
})(BodyTypes = exports.BodyTypes || (exports.BodyTypes = {}));
var ResponseMode;
(function (ResponseMode) {
    ResponseMode["RANDOM"] = "RANDOM";
    ResponseMode["SEQUENTIAL"] = "SEQUENTIAL";
    ResponseMode["DISABLE_RULES"] = "DISABLE_RULES";
})(ResponseMode = exports.ResponseMode || (exports.ResponseMode = {}));
var RouteType;
(function (RouteType) {
    RouteType["HTTP"] = "http";
    RouteType["CRUD"] = "crud";
})(RouteType = exports.RouteType || (exports.RouteType = {}));
var Methods;
(function (Methods) {
    Methods["get"] = "get";
    Methods["post"] = "post";
    Methods["put"] = "put";
    Methods["patch"] = "patch";
    Methods["delete"] = "delete";
    Methods["head"] = "head";
    Methods["options"] = "options";
    Methods["propfind"] = "propfind";
    Methods["proppatch"] = "proppatch";
    Methods["move"] = "move";
    Methods["copy"] = "copy";
    Methods["mkcol"] = "mkcol";
    Methods["lock"] = "lock";
    Methods["unlock"] = "unlock";
})(Methods = exports.Methods || (exports.Methods = {}));
