// @mockoon/commons@2.14.0/dist/esm/models/route.model.js
export var BodyTypes;
(function (BodyTypes) {
    BodyTypes["INLINE"] = "INLINE";
    BodyTypes["FILE"] = "FILE";
    BodyTypes["DATABUCKET"] = "DATABUCKET";
})(BodyTypes || (BodyTypes = {}));
export var ResponseMode;
(function (ResponseMode) {
    ResponseMode["RANDOM"] = "RANDOM";
    ResponseMode["SEQUENTIAL"] = "SEQUENTIAL";
    ResponseMode["DISABLE_RULES"] = "DISABLE_RULES";
})(ResponseMode || (ResponseMode = {}));
export var RouteType;
(function (RouteType) {
    RouteType["HTTP"] = "http";
    RouteType["CRUD"] = "crud";
})(RouteType || (RouteType = {}));
export var Methods;
(function (Methods) {
    Methods["get"] = "get";
    Methods["post"] = "post";
    Methods["put"] = "put";
    Methods["patch"] = "patch";
    Methods["delete"] = "delete";
    Methods["head"] = "head";
    Methods["options"] = "options";
    Methods["propfind"] = "propfind";
    Methods["proppatch"] = "proppatch";
    Methods["move"] = "move";
    Methods["copy"] = "copy";
    Methods["mkcol"] = "mkcol";
    Methods["lock"] = "lock";
    Methods["unlock"] = "unlock";
})(Methods || (Methods = {}));
