// @mockoon/commons@2.14.0/dist/cjs/constants/environment-schema.constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnvironmentSchema = exports.RouteSchema = exports.FolderSchema = exports.FolderChildSchema = exports.DataBucketDefault = exports.HeaderDefault = exports.ResponseRuleDefault = exports.RouteResponseDefault = exports.RouteDefault = exports.FolderDefault = exports.EnvironmentDefault = void 0;
const Joi = __importStar(require("joi"));
const uuid_1 = require("uuid");
const migrations_1 = require("../libs/migrations");
const utils_1 = require("../libs/utils");
const route_model_1 = require("../models/route.model");
exports.EnvironmentDefault = {
    get uuid() {
        return (0, uuid_1.v4)();
    },
    lastMigration: migrations_1.HighestMigrationId,
    name: 'New environment',
    endpointPrefix: '',
    latency: 0,
    port: 3000,
    hostname: '',
    folders: [],
    routes: [],
    rootChildren: [],
    proxyMode: false,
    proxyHost: '',
    proxyRemovePrefix: false,
    tlsOptions: {
        enabled: false,
        type: 'CERT',
        pfxPath: '',
        certPath: '',
        keyPath: '',
        caPath: '',
        passphrase: ''
    },
    cors: true,
    headers: [],
    proxyReqHeaders: [],
    proxyResHeaders: [],
    data: []
};
exports.FolderDefault = {
    get uuid() {
        return (0, uuid_1.v4)();
    },
    name: 'New folder',
    collapsed: false,
    children: []
};
exports.RouteDefault = {
    get uuid() {
        return (0, uuid_1.v4)();
    },
    type: route_model_1.RouteType.HTTP,
    documentation: '',
    method: route_model_1.Methods.get,
    endpoint: '',
    responses: [],
    enabled: true,
    responseMode: null
};
exports.RouteResponseDefault = {
    get uuid() {
        return (0, uuid_1.v4)();
    },
    body: '{}',
    latency: 0,
    statusCode: 200,
    label: '',
    headers: [],
    bodyType: route_model_1.BodyTypes.INLINE,
    filePath: '',
    databucketID: '',
    sendFileAsBody: false,
    rules: [],
    rulesOperator: 'OR',
    disableTemplating: false,
    fallbackTo404: false,
    default: false
};
exports.ResponseRuleDefault = {
    target: 'body',
    modifier: '',
    value: '',
    invert: false,
    operator: 'equals'
};
exports.HeaderDefault = {
    key: '',
    value: ''
};
exports.DataBucketDefault = {
    get uuid() {
        return (0, uuid_1.v4)();
    },
    get id() {
        return (0, utils_1.GenerateDatabucketID)();
    },
    name: 'New data',
    documentation: '',
    value: ''
};
const UUIDSchema = Joi.string()
    .uuid()
    .failover(() => (0, uuid_1.v4)())
    .required();
const HeaderSchema = Joi.object({
    key: Joi.string().allow('').required(),
    value: Joi.string().allow('').required()
});
const DataSchema = Joi.object({
    uuid: UUIDSchema,
    id: Joi.string().allow('').failover(exports.DataBucketDefault.id).required(),
    name: Joi.string().allow('').failover(exports.DataBucketDefault.name).required(),
    documentation: Joi.string()
        .allow('')
        .failover(exports.DataBucketDefault.documentation)
        .required(),
    value: Joi.string().allow('').failover(exports.DataBucketDefault.value).required()
})
    .failover(exports.EnvironmentDefault.data)
    .default(exports.EnvironmentDefault.data)
    .options({ stripUnknown: true });
const TLSOptionsSchema = Joi.object({
    enabled: Joi.boolean()
        .failover(exports.EnvironmentDefault.tlsOptions.enabled)
        .required(),
    type: Joi.string()
        .valid('CERT', 'PFX')
        .failover(exports.EnvironmentDefault.tlsOptions.type)
        .required(),
    pfxPath: Joi.string()
        .allow('')
        .failover(exports.EnvironmentDefault.tlsOptions.pfxPath)
        .required(),
    certPath: Joi.string()
        .allow('')
        .failover(exports.EnvironmentDefault.tlsOptions.certPath)
        .required(),
    keyPath: Joi.string()
        .allow('')
        .failover(exports.EnvironmentDefault.tlsOptions.keyPath)
        .required(),
    caPath: Joi.string()
        .allow('')
        .failover(exports.EnvironmentDefault.tlsOptions.caPath)
        .required(),
    passphrase: Joi.string()
        .allow('')
        .failover(exports.EnvironmentDefault.tlsOptions.passphrase)
        .required()
})
    .failover(exports.EnvironmentDefault.tlsOptions)
    .default(exports.EnvironmentDefault.tlsOptions)
    .options({ stripUnknown: true });
const RouteResponseRuleSchema = Joi.object({
    target: Joi.string()
        .valid('body', 'query', 'header', 'params', 'request_number', 'cookie')
        .failover(exports.ResponseRuleDefault.target)
        .required(),
    modifier: Joi.string()
        .allow('')
        .failover(exports.ResponseRuleDefault.modifier)
        .required(),
    value: Joi.string().allow('').failover(exports.ResponseRuleDefault.value).required(),
    invert: Joi.boolean().failover(exports.ResponseRuleDefault.invert).required(),
    operator: Joi.string()
        .valid('equals', 'regex', 'regex_i', 'null', 'empty_array')
        .failover(exports.ResponseRuleDefault.operator)
        .required()
});
const RouteResponseSchema = Joi.object({
    uuid: UUIDSchema,
    body: Joi.string().allow('').failover(exports.RouteResponseDefault.body).required(),
    latency: Joi.number()
        .min(0)
        .failover(exports.RouteResponseDefault.latency)
        .required(),
    statusCode: Joi.number()
        .min(100)
        .max(999)
        .failover(exports.RouteResponseDefault.statusCode)
        .required(),
    label: Joi.string().allow('').failover(exports.RouteResponseDefault.label).required(),
    headers: Joi.array()
        .items(HeaderSchema, Joi.any().strip())
        .failover(exports.RouteResponseDefault.headers)
        .required(),
    bodyType: Joi.string()
        .valid(route_model_1.BodyTypes.INLINE, route_model_1.BodyTypes.DATABUCKET, route_model_1.BodyTypes.FILE)
        .failover(exports.RouteResponseDefault.bodyType)
        .required(),
    filePath: Joi.string()
        .allow('')
        .failover(exports.RouteResponseDefault.filePath)
        .required(),
    databucketID: Joi.string()
        .allow('')
        .failover(exports.RouteResponseDefault.databucketID)
        .required(),
    sendFileAsBody: Joi.boolean()
        .failover(exports.RouteResponseDefault.sendFileAsBody)
        .required(),
    rules: Joi.array()
        .items(RouteResponseRuleSchema, Joi.any().strip())
        .failover(exports.RouteResponseDefault.rules)
        .required(),
    rulesOperator: Joi.string()
        .valid('OR', 'AND')
        .failover(exports.RouteResponseDefault.rulesOperator)
        .required(),
    disableTemplating: Joi.boolean()
        .failover(exports.RouteResponseDefault.disableTemplating)
        .required(),
    fallbackTo404: Joi.boolean()
        .failover(exports.RouteResponseDefault.fallbackTo404)
        .required(),
    default: Joi.boolean().failover(exports.RouteResponseDefault.default).required()
});
exports.FolderChildSchema = Joi.object({
    type: Joi.string()
        .valid('route', 'folder')
        .required()
        .options({ stripUnknown: true }),
    uuid: UUIDSchema
});
exports.FolderSchema = Joi.object({
    uuid: UUIDSchema,
    name: Joi.string().allow('').failover(exports.FolderDefault.name).required(),
    collapsed: Joi.boolean().failover(exports.FolderDefault.collapsed).required(),
    children: Joi.array()
        .items(exports.FolderChildSchema, Joi.any().strip())
        .failover(exports.FolderDefault.children)
        .required()
}).options({ stripUnknown: true });
exports.RouteSchema = Joi.object({
    uuid: UUIDSchema,
    type: Joi.string()
        .valid(route_model_1.RouteType.HTTP, route_model_1.RouteType.CRUD)
        .failover(exports.RouteDefault.type)
        .required(),
    documentation: Joi.string()
        .allow('')
        .failover(exports.RouteDefault.documentation)
        .required(),
    method: Joi.string()
        .allow('')
        .valid(route_model_1.Methods.get, route_model_1.Methods.post, route_model_1.Methods.put, route_model_1.Methods.patch, route_model_1.Methods.delete, route_model_1.Methods.head, route_model_1.Methods.options, route_model_1.Methods.propfind, route_model_1.Methods.proppatch, route_model_1.Methods.move, route_model_1.Methods.copy, route_model_1.Methods.mkcol, route_model_1.Methods.lock, route_model_1.Methods.unlock)
        .failover(exports.RouteDefault.method)
        .required(),
    endpoint: Joi.string().allow('').failover(exports.RouteDefault.endpoint).required(),
    responses: Joi.array()
        .items(RouteResponseSchema, Joi.any().strip())
        .failover(exports.RouteDefault.responses)
        .required(),
    enabled: Joi.boolean().failover(exports.RouteDefault.enabled).required(),
    responseMode: Joi.string()
        .allow(null)
        .valid(route_model_1.ResponseMode.RANDOM, route_model_1.ResponseMode.SEQUENTIAL, route_model_1.ResponseMode.DISABLE_RULES)
        .failover(exports.RouteDefault.responseMode)
        .required()
});
exports.EnvironmentSchema = Joi.object({
    uuid: UUIDSchema,
    lastMigration: Joi.number()
        .failover(exports.EnvironmentDefault.lastMigration)
        .required(),
    name: Joi.string().allow('').failover(exports.EnvironmentDefault.name).required(),
    endpointPrefix: Joi.string()
        .allow('')
        .failover(exports.EnvironmentDefault.endpointPrefix)
        .required(),
    latency: Joi.number().min(0).failover(exports.EnvironmentDefault.latency).required(),
    port: Joi.number()
        .min(0)
        .max(65535)
        .failover(exports.EnvironmentDefault.port)
        .required(),
    hostname: Joi.string()
        .allow('')
        .failover(exports.EnvironmentDefault.hostname)
        .required(),
    rootChildren: Joi.array()
        .items(exports.FolderChildSchema, Joi.any().strip())
        .failover(exports.EnvironmentDefault.rootChildren)
        .required(),
    folders: Joi.array()
        .items(exports.FolderSchema, Joi.any().strip())
        .failover(exports.EnvironmentDefault.folders)
        .required(),
    routes: Joi.array()
        .items(exports.RouteSchema, Joi.any().strip())
        .failover(exports.EnvironmentDefault.routes)
        .required(),
    proxyMode: Joi.boolean().failover(exports.EnvironmentDefault.proxyMode).required(),
    proxyHost: Joi.string()
        .allow('')
        .failover(exports.EnvironmentDefault.proxyHost)
        .required(),
    proxyRemovePrefix: Joi.boolean()
        .failover(exports.EnvironmentDefault.proxyRemovePrefix)
        .required(),
    tlsOptions: TLSOptionsSchema,
    cors: Joi.boolean().failover(exports.EnvironmentDefault.cors).required(),
    headers: Joi.array()
        .items(HeaderSchema, Joi.any().strip())
        .failover(exports.EnvironmentDefault.headers)
        .required(),
    proxyReqHeaders: Joi.array()
        .items(HeaderSchema, Joi.any().strip())
        .failover(exports.EnvironmentDefault.proxyReqHeaders)
        .required(),
    proxyResHeaders: Joi.array()
        .items(HeaderSchema, Joi.any().strip())
        .failover(exports.EnvironmentDefault.proxyResHeaders)
        .required(),
    data: Joi.array()
        .items(DataSchema, Joi.any().strip())
        .failover(exports.EnvironmentDefault.data)
        .required()
})
    .failover(exports.EnvironmentDefault)
    .default(exports.EnvironmentDefault)
    .options({ stripUnknown: true });
