// @mockoon/commons@2.14.0/dist/cjs/libs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.repairRefs = exports.GenerateDatabucketID = exports.RemoveLeadingSlash = exports.IsEqual = exports.CloneObject = exports.UnwrapLegacyExport = exports.IsLegacyExportData = exports.IsValidURL = exports.GetRouteResponseContentType = exports.GetContentType = void 0;
/**
 * Extract the content-type from an array of headers
 *
 * @param headers
 */
const GetContentType = (headers) => {
    const contentTypeHeader = headers.find((header) => header.key.toLowerCase() === 'content-type');
    if (contentTypeHeader) {
        return contentTypeHeader.value;
    }
    return null;
};
exports.GetContentType = GetContentType;
/**
 * Return a route response's content-type.
 * Environment's content-type is overridden by route's content-type
 *
 * @param environment
 * @param routeResponse
 */
const GetRouteResponseContentType = (environment, routeResponse) => {
    const routeResponseContentType = (0, exports.GetContentType)(routeResponse.headers);
    const environmentContentType = (0, exports.GetContentType)(environment.headers);
    return routeResponseContentType || environmentContentType || '';
};
exports.GetRouteResponseContentType = GetRouteResponseContentType;
/**
 * Test if URL is valid
 *
 * @param URL
 */
const IsValidURL = (address) => {
    try {
        const myURL = new URL(address);
        return true;
    }
    catch (e) {
        return false;
    }
};
exports.IsValidURL = IsValidURL;
/**
 * Verify if the object is a Mockoon legacy export data object
 *
 * @param data
 * @returns
 */
const IsLegacyExportData = (data) => {
    var _a;
    return data.source !== undefined &&
        ((_a = data.source) === null || _a === void 0 ? void 0 : _a.split(':')[0]) === 'mockoon';
};
exports.IsLegacyExportData = IsLegacyExportData;
/**
 * Import legacy export Mockoon's format.
 * Data was wrapped and could enclose multiple environments (and routes):
 *
 * ```
 * {
 *   "source": "mockoon:1.17.0",
 *   "data": [
 *     {
 *       "type": "environment",
 *       "item": {
 *         "uuid": "",
 *         "lastMigration": 13,
 *         "name": "Tutorial - Generate mock data"
 *         ...
 *       }
 *     }
 *   ]
 * }
 *
 * ```
 */
const UnwrapLegacyExport = (exportData) => exportData.data.reduce((environments, dataItem) => {
    if (dataItem.type === 'environment') {
        environments.push(dataItem.item);
    }
    return environments;
}, []);
exports.UnwrapLegacyExport = UnwrapLegacyExport;
/**
 * Clone an object using JSON.stringify
 * /!\ Suitable for Environment, Route, etc but not for complex objects containing Map, Set, etc
 */
const CloneObject = (objectToClone) => JSON.parse(JSON.stringify(objectToClone));
exports.CloneObject = CloneObject;
/**
 * Compare two objects using JSON.Stringify
 */
const IsEqual = (firstObject, secondObject) => JSON.stringify(firstObject) === JSON.stringify(secondObject);
exports.IsEqual = IsEqual;
const RemoveLeadingSlash = (str) => str.replace(/^\//g, '');
exports.RemoveLeadingSlash = RemoveLeadingSlash;
const GenerateDatabucketID = () => (Math.random() + 1).toString(36).substring(2, 6);
exports.GenerateDatabucketID = GenerateDatabucketID;
/**
 * Repair routes and folder references.
 * Remove references to non existing routes and folders.
 * Deduplicate references to the same route or folder.
 * Add references to orphan routes and folders at the root level.
 *
 * @param environment
 */
const repairRefs = (environment) => {
    const routesUUIDs = environment.routes.reduce((set, route) => {
        set.add(route.uuid);
        return set;
    }, new Set());
    const foldersUUIDs = environment.folders.reduce((set, folder) => {
        set.add(folder.uuid);
        return set;
    }, new Set());
    // remove folders children that are not existing
    environment.folders.forEach((folder) => {
        folder.children = folder.children.filter((folderChild) => {
            if (folderChild.type === 'route') {
                const hasItem = routesUUIDs.has(folderChild.uuid);
                if (hasItem) {
                    routesUUIDs.delete(folderChild.uuid);
                }
                return hasItem;
            }
            else {
                const hasItem = foldersUUIDs.has(folderChild.uuid);
                if (hasItem) {
                    foldersUUIDs.delete(folderChild.uuid);
                }
                return hasItem;
            }
        });
    });
    // remove root level children the are not existing
    environment.rootChildren = environment.rootChildren.filter((rootChild) => {
        if (rootChild.type === 'route') {
            const hasItem = routesUUIDs.has(rootChild.uuid);
            if (hasItem) {
                routesUUIDs.delete(rootChild.uuid);
            }
            return hasItem;
        }
        else {
            const hasItem = foldersUUIDs.has(rootChild.uuid);
            if (hasItem) {
                foldersUUIDs.delete(rootChild.uuid);
            }
            return hasItem;
        }
    });
    // add orphan folders to the root level
    foldersUUIDs.forEach((folderUUID) => {
        environment.rootChildren.push({
            type: 'folder',
            uuid: folderUUID
        });
    });
    // add orphan routes to the root level
    routesUUIDs.forEach((routeUUID) => {
        environment.rootChildren.push({
            type: 'route',
            uuid: routeUUID
        });
    });
    return environment;
};
exports.repairRefs = repairRefs;
